### PR TITLE
feat(ingest): Add helpers for fixing stuck project counters

### PR DESCRIPTION
**Pull request by lobsterkatie**
This adds two heper functions, `_is_stuck_counter_error` and  `_handle_stuck_project_counter`. The former is just a check on an error's message to see if it matches the known stuck counter errors. The latter corrects a project's project counter when it's gotten out of alignment with the project's highest `group.short_id` value and provides a new `short_id` value to use in group creation. These will be used in an upcoming PR to handle cases where a project counter has gotten stuck.

Note: In the interest of time, this PR doesn't include any unit tests for these functions. That said, the aforementioned upcoming PR will include integration tests using them, so they will get coverage then.

### CODE REVIEW

File: src/sentry/event_manager.py [hunk 1]
```diff
@@ -1891,6 +1891,55 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
...
+        if substring in error_message:
+            return True
+
+    return False
+
+
+def _handle_stuck_project_counter(project: Project, current_short_id: int) -> int:
+    """
+    Sometimes, for reasons unknown, a project's `Counter` value falls behind its latest group id.
+    When that happens, that incorrect counter value leads us to try to create groups with ids which
```
Comment (wedamija) [inline src/sentry/event_manager.py:1912]: Nit: `groups with short ids`

File: src/sentry/event_manager.py [hunk 2]
```diff
@@ -1891,6 +1891,55 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
...
+    are already taken.
+
+    This handles that case by updating the counter's value to the latest group id, and then returns
```
Comment (wedamija) [inline src/sentry/event_manager.py:1915]: Nit: `latest group short id`

File: src/sentry/event_manager.py [hunk 3]
```diff
@@ -1891,6 +1891,55 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
...
+    """
+    new_short_id = current_short_id
+
+    # Ordinarily running max on this many rows would be prohibitively expensive, but a) this is
+    # a very rare case (< 20 ever that we know of), and b) project and short id are indexed
+    # together in order to enforce the unique constraint which got us here in the first place,
+    # so it's faster than it otherwise might be.
+    max_short_id_for_project = Group.objects.filter(project_id=project.id).aggregate(
+        Max("short_id")
+    )["short_id__max"]
```
Comment (wedamija) [inline src/sentry/event_manager.py:1926]: To some extent I wonder if we should just skip the query and arbitrarily increment by 5 or something, but I guess this is cheap/rare enough that it's not a big problem

**Review by wedamija — APPROVED** [review]

File: src/sentry/event_manager.py [hunk 4]
```diff
@@ -1891,6 +1891,55 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
...
+    correct_value = max_short_id_for_project + 1
+
+    if current_short_id != correct_value:
+        difference = correct_value - current_short_id
+        # `_get_next_short_id` corrects the `Counter` value before it returns the new short_id
+        new_short_id = _get_next_short_id(project, delta=difference)
+
+        logger.info(
+            "Fixed stuck counter value.", extra={"project": project.id, "difference": difference}
+        )
```
Comment (wedamija) [inline src/sentry/event_manager.py:1938]: I think we should write something to datadog here so we can easily alert on it and see how frequent it is (bump sampling to 100%)

File: src/sentry/event_manager.py [hunk 5]
```diff
@@ -1891,6 +1891,55 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
     )
 
 
+def _is_stuck_counter_error(err: Exception, project: Project, short_id: int) -> bool:
+    """Decide if this is `UniqueViolation` error on the `Group` table's project and short id values."""
+
+    error_message = err.args[0]
+
+    for substring in [
+        f"Key (project_id, short_id)=({project.id}, {short_id}) already exists.",
```
[the remark this answers is not in the document]
Comment (lobsterkatie) [reply src/sentry/event_manager.py:1900]: FWIW this is already a partial match. (The full error message includes all the SQL and a million `%s`s.)

I'm curious: Why does taking the parameters out feel better to you, given that we have their values easily at hand.

File: src/sentry/event_manager.py (hunk 1 above)
Comment (lobsterkatie) [reply src/sentry/event_manager.py:1912]: Good catch. I originally thought they would be the same value, but once I realized they weren't, I thought I'd changed everywhere, but apparently not! Thanks, I'll fix it.

File: src/sentry/event_manager.py (hunk 2 above)
Comment (lobsterkatie) [reply src/sentry/event_manager.py:1915]: Also here.

File: src/sentry/event_manager.py (hunk 4 above)
Comment (lobsterkatie) [reply src/sentry/event_manager.py:1938]: It'll have to be in addition rather than instead, yeah? Since we can't add the project id to the metric?

For now I'll go with that.

Comment (wedamija) [reply src/sentry/event_manager.py:1939]: Yeah, both is fine

File: src/sentry/event_manager.py [hunk 6]
```diff
@@ -1891,6 +1891,59 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
...
+        new_short_id = _get_next_short_id(project, delta=difference)
+
+        logger.info(
+            "Fixed stuck counter value.", extra={"project": project.id, "difference": difference}
+        )
+        metrics.incr(
+            "stuck_project.fixed_counter", tags={"difference": difference}, sample_rate=1.0
+        )
+
+    return new_short_id
```
Comment (vartec) [inline src/sentry/event_manager.py:1944]: This returns a static integer, so there's a change of a race condition, i.e. when you use it to update Counter another group was create concurrently and it's invalid number again.

IMO the approach should be this returning a [SubQuery expression](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#subquery-expressions) with an [F() expression](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#f-expressions) to add 1.

In fact I think that using [OuterRef](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#django.db.models.OuterRef) this could be a completely static expression.

Basically that the resulting SQL would be something like

```sql
UPDATE sentry_projectcounter 
SET value = (
   SELECT max(short_id) + 1 
   FROM sentry_groupedmessage
   WHERE sentry_groupedmessage.project_id = sentry_projectcounter.project_id
)
WHERE sentry_projectcounter.project_id = %
```

File: src/sentry/event_manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/event_manager.py:1900]: Mostly just feels like extra complexity and doesn't buy us too much extra. It's a nit though, so if you prefer it this way I don't mind

File: src/sentry/event_manager.py (hunk 6 above)
Comment (wedamija) [reply src/sentry/event_manager.py:1944]: The way project counter works is to just increment by a delta. So the race condition here is that two groups try to fix this at the same time, and end up incrementing it multiple times

File: src/sentry/event_manager.py [hunk 7]
```diff
@@ -1891,6 +1891,59 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
...
+            Max("short_id")
+        )["short_id__max"]
+
+    # Add 1 because we're trying to mimic a value which would already have been incremented
+    correct_value = max_short_id_for_project + 1
+
+    if current_short_id != correct_value:
+        difference = correct_value - current_short_id
+        # `_get_next_short_id` corrects the `Counter` value before it returns the new short_id
+        new_short_id = _get_next_short_id(project, delta=difference)
```
Comment (wedamija) [inline src/sentry/event_manager.py:1935]: Let's make sure that there's no way this can be negative

File: src/sentry/event_manager.py (hunk 6 above)
Comment (lobsterkatie) [reply src/sentry/event_manager.py:1944]: > IMO the approach should be this returning a [SubQuery expression](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#subquery-expressions) with an [F() expression](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#f-expressions) to add 1.
> 
> In fact I think that using [OuterRef](https://docs.djangoproject.com/en/5.0/ref/models/expressions/#django.db.models.OuterRef) this could be a completely static expression.
> 
> Basically that the resulting SQL would be something like
> 
> ```sql
> UPDATE sentry_projectcounter 
> SET value = (
>    SELECT max(short_id) + 1 
>    FROM sentry_groupedmessage
>    WHERE sentry_groupedmessage.project_id = sentry_projectcounter.project_id
> )
> WHERE sentry_projectcounter.project_id = %
> ```

Interesting. I can look into those. In the meantime, though: 

> This returns a static integer, so there's a change of a race condition, i.e. when you use it to update Counter another group was create concurrently and it's invalid number again.

I get what you're saying, but whouldn't that apply just as much to regular group creation? Two groups come in at the same time, both call `project.next_short_id`, and end up with the same short_id? Under the hood, that's all this really is - another call to `next_short_id`.

That said, @wedamija and I actually did have a conversation the other day about locks here. I was suggesting possibly something like we do [when we create a new group based on has](https://github.com/getsentry/sentry/blob/2e4feb99bc3c6c44d7abe438a23e1e2ee4c7e2ce/src/sentry/event_manager.py#L1812-L1836)h, but then he mentioned we might be able to do it with redis. Dan, thoughts?

File: src/sentry/event_manager.py (hunk 7 above)
Comment (lobsterkatie) [reply src/sentry/event_manager.py:1935]: Dangit. I had the check here as `current_short_id < correct_value` until I was writing the PR description and looked at it and thought, maybe I should make this more general. 😓 

I can switch it back.

File: src/sentry/event_manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vartec) [reply src/sentry/event_manager.py:1900]: optimization: before getting to searching for substrings, fail fast with:
```python  
if not error_message.startswith('UniqueViolation'):
   return False
```

File: src/sentry/event_manager.py (hunk 6 above)
Comment (vartec) [reply src/sentry/event_manager.py:1944]: > two groups try to fix this at the same time, and end up incrementing it multiple times

True, on the other hand if two groups run the update with subselect query at the same time it should be all good, right?

Comment (wedamija) [reply src/sentry/event_manager.py]: No, the groups don't exist yet, they are about to be created. So two new groups run the query at the same time and end up with the same short id, then one fails. Unless we get them to both call increment after

Comment (vartec) [reply src/sentry/event_manager.py]: @lobsterkatie 
> I get what you're saying, but whouldn't that apply just as much to regular group creation? Two groups come in at the same time, both call project.next_short_id, and end up with the same short_id? Under the hood, that's all this really is - another call to next_short_id.

More like the counter is `1` behind, two groups are created concurrently, both call `_handle_stuck_project_counter`, both calculate delta to be `3`.

Comment (lobsterkatie) [reply src/sentry/event_manager.py]: > both call _handle_stuck_project_counter, both calculate delta to be 3

Right, I get the setup. But two simultaneous calls to `next_short_id` are two simultaneous calls to `next_short_id`, no? That can happen in normal projects when two groups try to be created at the same time, or in a stuck project when two groups try to fix the counter at the same time, or in a stuck project where one group calls it at the top of `_create_group` and one calls it in this function, or when both call it at the top of `_create_group`. It seems like if one of these is a problem, they're all a problem, but the first scenario (two normal groups) happens without incident many times a day. We must work that out somehow...

File: src/sentry/event_manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lobsterkatie) [reply src/sentry/event_manager.py:1903]: Good idea. Done.

File: src/sentry/event_manager.py (hunk 6 above)
Comment (lobsterkatie) [reply src/sentry/event_manager.py]: UPDATE: Based on my conversation with @wedamija [here](https://github.com/getsentry/sentry/pull/65655#discussion_r1500080724), I'm going to go ahead and merge these PRs and we can cross the race condition problem if it comes up.
